src/System.Collections/tests/Generic/HashSet/HashSet.Generic.Tests.cs

```diff
@@ -162,6 +162,122 @@ public void HashSet_Generic_RemoveWhere_NullMatchPredicate(int setLength)
 
         #endregion
 
+        #region EnsureCapacity
+        
+        [Theory]
+        [MemberData(nameof(ValidCollectionSizes))]
+        public void EnsureCapacity_Generic_RequestingLargerCapacity_DoesNotInvalidateEnumeration(int setLength)
+        {
+            HashSet<T> set = (HashSet<T>)(GenericISetFactory(setLength));
+            var capacity = set.EnsureCapacity(0);
+            IEnumerator<T> valuesEnum = set.GetEnumerator();
+            set.EnsureCapacity(capacity + 1); // Verify EnsureCapacity does not invalidate enumeration
+            while (valuesEnum.MoveNext());
+        }
+
+        [Fact]
+        public void EnsureCapacity_Generic_NegativeCapacityRequested_Throws()
+        {
+            var set = new HashSet<T>();
+            AssertExtensions.Throws<ArgumentOutOfRangeException>("capacity", () => set.EnsureCapacity(-1));
+        }
+
+        [Fact]
+        public void EnsureCapacity_Generic_HashsetNotInitialized_RequestedZero_ReturnsZero()
+        {
+            var set = new HashSet<T>();
+            Assert.Equal(0, set.EnsureCapacity(0));
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(2)]
+        [InlineData(3)]
+        [InlineData(4)]
+        public void EnsureCapacity_Generic_HashsetNotInitialized_RequestedNonZero_CapacityIsSetToAtLeastTheRequested(int requestedCapacity)
+        {
+            var set = new HashSet<T>();
+            Assert.InRange(set.EnsureCapacity(requestedCapacity), requestedCapacity, int.MaxValue);
+        }
+
+        [Theory]
+        [InlineData(3)]
+        [InlineData(7)]
+        public void EnsureCapacity_Generic_RequestedCapacitySmallerThanCurrent_CapacityUnchanged(int currentCapacity)
+        {
+            HashSet<T> set;
+
+            // assert capacity remains the same when ensuring a capacity smaller or equal than existing
+            for (int i = 0; i <= currentCapacity; i++)
+            {
+                set = new HashSet<T>(currentCapacity);
+                Assert.Equal(currentCapacity, set.EnsureCapacity(i));
+            }
+        }
+
+        [Theory]
+        [InlineData(7)]
+        [InlineData(89)]
+        public void EnsureCapacity_Generic_ExistingCapacityRequested_SameValueReturned(int capacity)
+        {
+            var set = new HashSet<T>(capacity);
+            Assert.Equal(capacity, set.EnsureCapacity(capacity));
+
+            set = (HashSet<T>)GenericISetFactory(capacity);
+            Assert.Equal(capacity, set.EnsureCapacity(capacity));
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(7)]
+        public void EnsureCapacity_Generic_EnsureCapacityCalledTwice_ReturnsSameValue(int setLength)
+        {
+            HashSet<T> set = (HashSet<T>)GenericISetFactory(setLength);
+            int capacity = set.EnsureCapacity(setLength * 2);
```

Comment:
Does * 2 change anything? It will still be a constant value, it will be 14 instead of 7 and 0 will remain the same. Having it just kind of confuses the purpose of the test, it seems like if you where increasing the value every call.

Reply:
I agree it's confusing, I'll improve in the next commit.